### PR TITLE
Transform result

### PR DESCRIPTION
**Pull request by EasyRhin0**
Since a result can have multiple locations but only one analysisTarget, I am just grabbing the analysisTarget from location[0].

### CODE REVIEW

**Comment by lgolding** [comment]
That is correct.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 1]
```diff
@@ -953,5 +951,168 @@ public void SarifTransformerTests_ToCurrent_NotificationExceptionWithStack()
...
+                }
+              },
+              ""fullyQualifiedLogicalName"": ""collections::list::add""
+            }
+          ],
+          ""suppressionStates"": [""suppressedExternally""],
+          ""baselineState"": ""existing"",
+          ""properties"": {
+            ""sarifv1/formattedRuleMessage"": {""formatId"":""default"",""arguments"":[""ptr""]},
+            ""sarifv1/snippet"": ""add_core(ptr, offset, val);\n    return;""
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: You can put this in `region.snippet` of type `fileContents`.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 2]
```diff
@@ -114,6 +114,22 @@ internal FileLocation CreateFileLocation(Uri uri, string uriBaseId)
...
+            if (v1PhysicalLocation?.Uri != null)
+            {
+                fileLocation = new FileLocation
+                {
+                    Uri = v1PhysicalLocation.Uri,
+                    UriBaseId = v1PhysicalLocation.UriBaseId
+                };
+            }
+
+            return fileLocation;
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Doesn't this exist somewhere else?

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 3]
```diff
@@ -190,6 +206,28 @@ internal Invocation CreateInvocation(InvocationVersionOne v1Invocation)
...
+                {
+                    FullyQualifiedLogicalName = v1Location.LogicalLocationKey ?? v1Location.FullyQualifiedLogicalName,
+                    PhysicalLocation = CreatePhysicalLocation(v1Location.ResultFile),
+                    Properties = v1Location.Properties
+                };
+
+                if (_currentRun.LogicalLocations.ContainsKey(location.FullyQualifiedLogicalName))
+                {
+                    _currentRun.LogicalLocations[location.FullyQualifiedLogicalName].DecoratedName = v1Location.DecoratedName;
+                }
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: If the entry doesn't exist, we should create it, right?

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 2 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: There is an overload that takes the uri and uribaseid

Comment (lgolding) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: ok.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: When you do that, let's make the properties of the region consistent with the length of the snippet.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 3 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: The logicalLocations dictionary has been imported at this point, wouldn't that definitely include it?

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: or locations[0] maybe

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 3 above)
Comment (lgolding) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Not necessarily.  `run.logicalLocations` is optional (in both v1 and v2). Now, there's nowhere else to put `decoratedName`, so if you have `decoratedName`, you have to make sure `run.logicalLocations` is present. Come to think of it, you need to check for the existence of the dictionary before adding to it. And I suspect you'll be doing that check-for-dictionary-and-create-if-necessary-and-then-add-entry multiple times, so you might want to factor it out.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Put the snippet in `locations[0].physicalLocation.region.snippet`, and make sure the properties of the `region` in your test log are consistent with the text of the `snippet`.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 3 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Yeah I already have AddLogicalLocation which should work here.

Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Done

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 4]
```diff
@@ -547,7 +553,19 @@ internal Result CreateResult(ResultVersionOne v1Result)
...
+                    {
+                        region.Snippet = new FileContent
+                        {
+                            Text = v1Result.Snippet
+                        };
+
+                        region.Length = v1Result.Snippet.Length;
+                        region.StartColumn = 1;
+                        region.EndColumn = region.Length + 1;
+                    }
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: I'm sorry, I was unclear. I meant: "For the sake of the people who read the v1 sample file, make the region properties in the v1 sample file consistent with the snippet in the v1 sample file." When it comes to actually _copying_ the properties from v1 to v2, just take what they give you. If the v1 file was inconsistent, tough luck. (In fact, you _have_ to just copy the properties, for the sake of round-tripping (unless and until we adopt the "stuff the whole v1 run in the v2 run's property bag" trick).

Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: ok sorry, should have known that :p

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 5]
```diff
@@ -467,6 +511,63 @@ internal Region CreateRegion(int startColumn, int startLine, int endColumn = 0,
...
+                {
+                    // Only the formattedRuleMessage.formatId property is used above
+                    result.SetProperty($"{FromPropertyBagPrefix}/formattedRuleMessage", v1Result.FormattedRuleMessage);
+                }
+
+                if (!string.IsNullOrWhiteSpace(v1Result.Snippet))
+                {
+                    Region region = result.Locations?[0]?.PhysicalLocation?.Region;
+
+                    if (region != null)
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: We have to store the snippet. If the region isn't there, create it. Likewise for its ancestors the physicalLocation object, the location object, and the locations array.

**Pull request merged**